File: core/src/main/scala/diplomacy/variant/standard/rule/BuildResolver.scala
package diplomacy.variant.standard.rule

import scala.collection.{ mutable => cmutable }

import diplomacy.rule.{ InvalidOrderMessage, OrderResult, ResolvedResult }
import diplomacy.variant.standard.rule.Keywords._

class BuildResolver extends Rule.TypeHelper {
  def apply(nextTurn: Turn => Turn)(
    board: Board, orders: Set[Order]
  ): Either[InvalidOrderMessage, ResolvedResult] = {
    val disbands = orders collect { case d@ Order.Disband(_) => d }
    val builds = orders collect { case b@ Order.Build(_) => b }

    val newUnits = board.units -- (disbands map { _.unit }) ++ (builds map { _.unit })
    val newState =
      board.state.copy(turn = nextTurn(board.state.turn), phase = Movement)

    val newBoard = board.copy(
      state = newState,
      units = newUnits,
      unitStatuses = Map[DiplomacyUnit, UnitStatus](),
      provinceStatuses = Map[Province, ProvinceStatus]()
    )
    val orderResults: Set[OrderResult] = ((disbands ++ builds) map {
      case order =>
        OrderResult.Executed[Power, MilitaryBranch, Order, Result](order, Result.Success)
    })(collection.breakOut)

    Right(ResolvedResult(newBoard, orderResults, false))
  }
}

File: lib/variant/standard/rule/build-resolver.js
const { Board } = require('../../../board/package')
const { OrderResult, ResolvedResult } = require('../../../rule/package')
const State = require('./state')
const ProvinceStatus = require('./province-status')
const { Movement } = require('./phase')
const { Success } = require('./result')

module.exports = class BuildResolver {
  resolve (board, orders) {
    const disbands = [...orders].filter(order => order.tpe === 'Disband')
    const builds = [...orders].filter(order => order.tpe === 'Build')

    const newUnits = new Set([...board.units])
    disbands.forEach(d => newUnits.delete(d.unit))
    builds.forEach(b => newUnits.add(b.unit))

    const newState = new State(board.state.turn.nextTurn(), Movement)
    const occupationStatuses = [...board.provinceStatuses]
      .filter(elem => elem[1].occupied)
      .map(elem => {
        if (elem[1].standoff) {
          return [elem[0], new ProvinceStatus(elem[1].occupied, false)]
        } else {
          return elem
        }
      })
    const newBoard = new Board(board.map, newState, newUnits, [], occupationStatuses)
    const orderResults = [...orders].map(order => new OrderResult.Executed(order, Success))

    return { result: new ResolvedResult(newBoard, orderResults, false) }
  }
}
